baseplate/secrets/fetcher.py

```diff
@@ -0,0 +1,282 @@
+
+            vault write /auth/aws-ec2/role/my-server-role \
+                bound_iam_arn=arn:aws:iam::12341234:role/my-servers-iam-role \
+                policies=my-servers-policies \
+                max_ttl=4h
+
+        To combat replay attacks where the identity document is snarfed and
+        passed on by an interloper, during first login we store a
+        Vault-generated nonce locally in a protected file. This nonce must be
+        passed back to Vault on all successive login attempts.
+
+        See https://www.vaultproject.io/docs/auth/aws-ec2.html for more info.
+
+        """
+        identity_document = fetch_instance_identity()
+        nonce = load_nonce()
+
+        login_data = {
+            "role": self.role,
+            "pkcs7": identity_document,
+        }
+
+        if nonce:
+            login_data["nonce"] = nonce
+
+        logger.debug("Logging into Vault.")
+        response = self.session.post(
+            urljoin(self.base_url, "/v1/auth/aws-ec2/login"),
+            json=login_data,
+        )
+        response.raise_for_status()
+        response_payload = response.json()
+
+        nonce = response_payload["auth"]["metadata"]["nonce"]
+        store_nonce(nonce)
+
+        return VaultClient(
+            self.session,
+            self.base_url,
+            response_payload["auth"]["client_token"],
+            ttl_to_time(response_payload["auth"]["lease_duration"]),
+        )
+
+    def get_client(self):
+        """Get an authenticated client, reauthenticating if not cached."""
+        if not self.client or self.client.is_about_to_expire:
+            self.client = self._make_client()
+        return self.client
+
+
+class VaultClient(object):
+    """An authenticated vault client.
+
+    Use this as long as is_about_to_expire is False. Once that becomes True,
+    get a new client from the factory.
+
+    """
+    def __init__(self, session, base_url, token, token_expiration):
+        self.session = session
+        self.base_url = base_url
+        self.token = token
+        self.token_expiration = token_expiration
+
+    @property
+    def is_about_to_expire(self):
+        """Is the token near expiration and in need of regeneration?"""
+        expiration = self.token_expiration - VAULT_TOKEN_PREFETCH_TIME
+        return expiration < datetime.datetime.utcnow()
+
+    def get_secret(self, secret_name):
+        """Get the value and expiration time of a named secret."""
+        logger.debug("Fetching secret %r.", secret_name)
+        response = self.session.get(
+            urljoin(self.base_url, posixpath.join("v1", secret_name)),
+            headers={
+                "X-Vault-Token": self.token,
+            },
+        )
+        response.raise_for_status()
+        payload = response.json()
+        return payload["data"], ttl_to_time(payload["lease_duration"])
+
+
+def octal_integer(text):
+    return int(text, base=8)
+
+
+def user_name(text):
```

Comment:
Hm, yeah, it looks a bit weird when doing it like this but I'm going for the declarative style of the rest of the config helpers. Maybe I should capitalize like they do? or just ditch it?